Orbitersdk/samples/ProjectApollo/src_sys/inertial.cpp

```diff
@@ -0,0 +1,181 @@
+
+**************************************************************************/
+
+#include "inertial.h"
+#include "papi.h"
+
+InertialData::InertialData(VESSEL *vessel) {
+	this->vessel = vessel;
+	accel = _V(0.0, 0.0, 0.0);
+}
+
+VECTOR3 InertialData::GetGravityVector() {
+	OBJHANDLE gravref = vessel->GetGravityRef();
+	OBJHANDLE hSun = oapiGetObjectByName("Sun");
+	VECTOR3 R, U_R;
+	vessel->GetRelativePos(gravref, R);
+	U_R = unit(R);
+	double r = length(R);
+	VECTOR3 R_S, U_R_S;
+	vessel->GetRelativePos(hSun, R_S);
+	U_R_S = unit(R_S);
+	double r_S = length(R_S);
+	double mu = GGRAV * oapiGetMass(gravref);
+	double mu_S = GGRAV * oapiGetMass(hSun);
+	int jcount = oapiGetPlanetJCoeffCount(gravref);
+	double JCoeff[5];
+	for (int i = 0; i < jcount; i++)
+	{
+		JCoeff[i] = oapiGetPlanetJCoeff(gravref, i);
+	}
+	double R_E = oapiGetSize(gravref);
+
+	VECTOR3 a_dP;
+
+	a_dP = -U_R;
+
+	if (jcount > 0)
+	{
+		MATRIX3 mat;
+		VECTOR3 U_Z;
+		double costheta, P2, P3;
+
+		oapiGetPlanetObliquityMatrix(gravref, &mat);
+		U_Z = mul(mat, _V(0, 1, 0));
+
+		costheta = dotp(U_R, U_Z);
+
+		P2 = 3.0 * costheta;
+		P3 = 0.5*(15.0*costheta*costheta - 3.0);
+		a_dP += (U_R*P3 - U_Z * P2)*JCoeff[0] * pow(R_E / r, 2.0);
+		if (jcount > 1)
+		{
+			double P4;
+			P4 = 1.0 / 3.0*(7.0*costheta*P3 - 4.0*P2);
+			a_dP += (U_R*P4 - U_Z * P3)*JCoeff[1] * pow(R_E / r, 3.0);
+			if (jcount > 2)
+			{
+				double P5;
+				P5 = 0.25*(9.0*costheta*P4 - 5.0 * P3);
+				a_dP += (U_R*P5 - U_Z * P4)*JCoeff[2] * pow(R_E / r, 4.0);
+			}
+		}
+	}
+	a_dP *= mu / pow(r, 2.0);
+	a_dP -= U_R_S * mu_S / pow(r_S, 2.0);
+
+	if (gravref == oapiGetObjectByName("Moon"))
+	{
+		OBJHANDLE hEarth = oapiGetObjectByName("Earth");
+
+		VECTOR3 R_Ea, U_R_E;
+		vessel->GetRelativePos(hEarth, R_Ea);
+		U_R_E = unit(R_Ea);
+		double r_E = length(R_Ea);
+		double mu_E = GGRAV * oapiGetMass(hEarth);
+
+		a_dP -= U_R_E * mu_E / pow(r_E, 2.0);
+	}
+
+	return a_dP;
+}
+
+
+void InertialData::timestep(double simdt) {
+	VECTOR3 w, vel;
+
+	vessel->GetWeightVector(w);
+	vessel->GetGlobalVel(vel);
+	vessel->GetRotationMatrix(rotmat);
+
+	w = mul(rotmat, w) / vessel->GetMass();
+
+	//Orbiter 2016 hack
+	if (length(w) == 0.0)
+	{
+		w = GetGravityVector();
+	}
+
+	if (!dVInitialized) {
+		lastWeight = w;
+		lastGlobalVel = vel;
+		lastSimDT = simdt;
+		dVInitialized = true;
+	}
+	else {
+		// Acceleration calculation, see IMU
+		VECTOR3 dvel = (vel - lastGlobalVel) / lastSimDT;
+		VECTOR3 dw1 = w - dvel;
+		VECTOR3 dw2 = lastWeight - dvel;
+		lastWeight = w;
+		lastGlobalVel = vel;
+		lastSimDT = simdt;
+
+		// Transform to vessel coordinates
+		VECTOR3 avg = (dw1 + dw2) / 2.0;
+		accel = tmul(rotmat, avg);
+	}
+}
+
+void InertialData::getAcceleration(VECTOR3 &acc) {
+	acc.x = accel.x;
+	acc.y = accel.y;
+	acc.z = accel.z;
+}
+
+void InertialData::SaveState(FILEHANDLE scn){
+	oapiWriteLine(scn, INERTIAL_DATA_START_STRING);
+	oapiWriteScenario_int(scn, "DVINITIALIZED", (dVInitialized ? 1 : 0));
```

Comment:
I know this was already saved in the previous implementation. However, would not saving these and instead reinitializing them on each load cause any problems? It would remove four variables from the ever increasing list of things to save.

Reply:
Maybe there is a way, but I think the problem is that, despite the timestep function being called in clbkPostStep, in Orbiter 2016 the simulation has changed so that functions like GetWeightVector aren't updated until AFTER clbkPostStep, that was one of the first issues to fix in NASSP when we moved to 2016. See this thread: https://www.orbiter-forum.com/threads/vessel-getglobalvel-and-or-simdt-oscillation.34756/#post-533107

So when this timestep function is being called for the first time after starting the sim, the weight vector etc. are still identical to the values that Orbiter would return during your proposed reinitialization function call, which would be during scenario loading. So I don't really see a way to do it unless we change how our accelerometers work in general.

Also, this PR will actually reduce the number of scenario lines, because each accelerometer class (IMU etc.) won't need to save this data anymore. Just look how many oapiWriteScenario/papiWriteScenario are being deleted!

Reply:
Sounds reasonable. Just making sure it can't be optimized away all together. Let's keep it then.